lib/websockex.ex

```diff
@@ -1123,6 +1129,33 @@ defmodule WebSockex do
     end
   end
 
+  defp handle_send_result(result, frame, key, state) do
+    result = try_callback(state.module, :handle_send_result, [result, frame, key, state.module_state])
+
+    case result do
+      {:ok, new_state} ->
+        {:ok, new_state}
+
+      {:close, new_state} ->
+        {:close, new_state}
+
+      {:close, close_frame, new_state} ->
+        {:close, close_frame, new_state}
+
+      {:"$EXIT", _} = res ->
+        res
+
+      badreply ->
+        {:"$EXIT",
+         %WebSockex.BadResponseError{
+           module: state.module,
+           function: :handle_send_result,
+           args: [result, frame, state.module_state],
+           response: badreply
+         }}
+    end
+  end
```

Comment:
I understand the desire for this, but I think that this is actually a not great idea.

TLDR; Just because a message is sent doesn't mean that it was received. Rebuilding messages that need to be sent should depend on what the server has processed.

The reason why is that the WebSocket protocol actually doesn't have an acknowledgement mechanism built into it. Most people build those mechanisms around the protocol, erroring out if processing a message out of order, etc. We have a guarantee that our message was sent because of TCP guarantees, but we have no idea if our message was processed without an acknowledgement.

Some servers do this with custom ping messages wrapped in data frames, some servers require unsolicited pong messages with data payloads, some servers only use WebSockets for unidirectional data. There's a whole world out there of ways of handling messages that were sent.

But because the underlying mechanism is TCP, you could send 5 additional frames before the server responds that it is unable to process a frame. We may be able to say that this frame "failed to send", but the current frame wasn't the problem. Likewise, your last frame could have succeeded in getting to the proxy, but it never made it to the server.

So, I think that offering this kind of functionality would actually lead to bad practices in the design of the connection itself. Deciding on what needs to be sent or resent on a connection should be determined after getting some state from the server.

Reply:
Ye I understand there are many network failures that could occur. I actually added this because adding a backoff state changes the behaviour quite a bit. For example, if exponential backoff is implemented, it's not unrealistic that someone could set a max backoff to 10s depending on their application/use case. But imo this still applies even for a backoff of 1s.

On the one hand, we don't want to prevent `handle_cast` messages from being processed while in the backoff state (ie. excluding `handle_cast` messages from the selective receive in backoff state).
On the other hand, doing a `:reply` while in the backoff state results in a 100% chance for failure, which can be acted upon immediately by e.g. stopping sending more messages until connected again, or in my use case, NACKing the message immediately (see https://github.com/jeffgrunewald/stargate/pull/26). The stopping sending messages until connected can of course be implemented via the `handle_connect` and `handle_disconnect` callbacks, but there's still the chance of messages getting lost because of race conditions, which can then only be NACK'ed on a timeout.

So yes this isn't a replacement for application ACKs, but rather a way to give fast feedback of messages guaranteed to NOT be delivered during a backoff state.

Reply:
You're not wrong, but after reading your explanation I feel the deficiency in the WebSockex API may be that we need to have some kind of connection state passed to `handle_cast` and `handle_info`.

What are your thoughts on that?

Reply:
I thought about this and decided against it because that solution still allows for race conditions. 
If in `handle_cast` you check that the connection is connected and then do a `reply`, and the connection drops before the `reply` is processed, then you still lose a message. Having a callback `handle_send_result` puts `reply` in the same position as `send_frame`, just that the handling is async. `send_frame` gives you feedback on whether the TCP packet was sent or if the connection wasn't established, e.g. by returning a `NotConnected` error. 
IMO having these 2 apis be equally powerful in terms of gaurantees/feedback it provides is good.

The other thing I thought is that  people have also asked for allowing the same features as `GenServer`. If `WebSockex` adds `handle_call` support then together with `handle_send_result`, a user is able to implement `send_frame` without it being a native api.

In general I feel that async apis should be more powerful, not less powerful than sync apis, and in any case, in Elixir, sync apis are always an abstraction over async apis.
So it just feels wrong that `reply` is less powerful/reliable than `send_frame` in terms of feedback/reliability.